File: backend/models/roster.py
from datetime import datetime, timezone
from typing import Literal
import uuid

from pydantic import BaseModel, Field, field_validator, model_validator


DayName = Literal["Senin", "Selasa", "Rabu", "Kamis", "Jumat", "Sabtu"]
ValidationLevel = Literal["ok", "warning", "error"]


class ClassSlot(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    day: DayName
    start_time: str
    end_time: str
    course: str

    @field_validator("start_time", "end_time")
    @classmethod
    def validate_time(cls, value: str) -> str:
        try:
            datetime.strptime(value, "%H:%M")
        except ValueError as exc:
            raise ValueError("waktu harus memakai format HH:MM") from exc
        return value

    @model_validator(mode="after")
    def validate_range(self):
        if self.start_time >= self.end_time:
            raise ValueError("jam selesai harus setelah jam mulai")
        return self


class MemberSchedule(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    code: str
    division: str
    generation: str = Field(pattern=r"^[0-9]{2}$")
    source_file: str
    confidence: float = Field(ge=0, le=1)
    notes: list[str] = Field(default_factory=list)
    classes: list[ClassSlot] = Field(default_factory=list)


class ExtractionResponse(BaseModel):
    members: list[MemberSchedule]
    warnings: list[str] = Field(default_factory=list)


class PlotConfig(BaseModel):
    shift_duration_hours: int = Field(default=2, ge=1, le=4)
    operating_start: str = "07:00"
    operating_end: str = "18:00"
    active_days: list[DayName] = Field(
        default_factory=lambda: ["Senin", "Selasa", "Rabu", "Kamis", "Jumat"]
    )

    @field_validator("operating_start", "operating_end")
    @classmethod
    def validate_time(cls, value: str) -> str:
        try:
            datetime.strptime(value, "%H:%M")
        except ValueError as exc:
            raise ValueError("waktu harus memakai format HH:MM") from exc
        return value

    @model_validator(mode="after")
    def validate_range(self):
        if self.operating_start >= self.operating_end:
            raise ValueError("jam operasional selesai harus setelah jam mulai")
        if not self.active_days:
            raise ValueError("minimal satu hari aktif diperlukan")
        return self


class PlotRequest(BaseModel):
    batch_id: str
    schedules: list[MemberSchedule] = Field(min_length=1)
    config: PlotConfig


class PlotAssignment(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    day: DayName
    start_time: str
    end_time: str
    member_code: str
    division: str
    generation: str


class ValidationItem(BaseModel):
    level: ValidationLevel
    day: DayName | None = None
    message: str


class PlotResponse(BaseModel):
    assignments: list[PlotAssignment]
    validations: list[ValidationItem]
    divisions: list[str]
    required_divisions: list[str]
    missing_divisions: list[str]
    coverage_by_day: dict[str, list[str]]
    complete_days: list[DayName]
    export_ready: bool
    generated_at: str = Field(default_factory=lambda: datetime.now(timezone.utc).isoformat())


class ExportRequest(BaseModel):
    batch_id: str
    plot: PlotResponse

File: frontend/src/types/roster.ts
export type DayName = "Senin" | "Selasa" | "Rabu" | "Kamis" | "Jumat" | "Sabtu";
export type ValidationLevel = "ok" | "warning" | "error";

export interface ClassSlot {
  id: string;
  day: DayName;
  start_time: string;
  end_time: string;
  course: string;
}

export interface MemberSchedule {
  id: string;
  code: string;
  division: string;
  generation: string;
  source_file: string;
  confidence: number;
  notes: string[];
  classes: ClassSlot[];
}

export interface ExtractionResponse {
  members: MemberSchedule[];
  warnings: string[];
}

export interface PlotConfig {
  shift_duration_hours: number;
  operating_start: string;
  operating_end: string;
  active_days: DayName[];
}

export interface PlotRequest {
  batch_id: string;
  schedules: MemberSchedule[];
  config: PlotConfig;
}

export interface PlotAssignment {
  id: string;
  day: DayName;
  start_time: string;
  end_time: string;
  member_code: string;
  division: string;
  generation: string;
}

export interface ValidationItem {
  level: ValidationLevel;
  day: DayName | null;
  message: string;
}

export interface PlotResponse {
  assignments: PlotAssignment[];
  validations: ValidationItem[];
  divisions: string[];
  required_divisions: string[];
  missing_divisions: string[];
  coverage_by_day: Record<string, string[]>;
  complete_days: DayName[];
  export_ready: boolean;
  generated_at: string;
}

export interface ExportRequest {
  batch_id: string;
  plot: PlotResponse;
}

export type QueueStatus = "valid" | "invalid" | "unsupported" | "duplicate" | "processing" | "done" | "error" | "cached";

export interface ParsedFilename {
  code: string;
  division: string;
  generation: string;
  extension?: string;
  supported: boolean;
  valid: boolean;
  error?: string;
}

export interface QueuedFile {
  id: string;
  file: File;
  hash: string;
  parsed: ParsedFilename;
  status: QueueStatus;
  error?: string;
  duplicateOf?: string;
  cachedMembers?: MemberSchedule[];
}

export interface BatchHistory {
  id: string;
  name: string;
  created_at: string;
  updated_at: string;
  members: MemberSchedule[];
  config: PlotConfig;
  plot: PlotResponse | null;
  source_hashes: string[];
  reviewed: boolean;
  xlsx?: Blob;
}

export interface ExtractionCacheEntry {
  hash: string;
  members: MemberSchedule[];
  updated_at: string;
}